### PR TITLE
Implement VSC AST using python dataclasses 

### PR DESCRIPTION
**Pull request by miketsukerman**
**Vehicle Service Catalog AST using python dataclasses**

The general idea is to use https://docs.python.org/3/library/dataclasses.html instead of a python dictionary to describe the types from which an AST would consist. 

**Example code**

Example python code for the `method` element. (https://github.com/COVESA/vehicle_service_catalog/blob/master/syntax.md#namespace-list-object-methods)

```yaml
methods:
    - name: move
      description: Set the desired seat position
      input: 
         - name: seat
           description: The desired seat position
           datatype: movement.seat_t
```

will correspond to the following python code:

```python
@dataclass
class Method:
    name: str
    description: Optional[str] = None
    error: Optional[List[Error]] = field(default_factory=EmptyList)
    input: Optional[List[Argument]] = field(default_factory=EmptyList)
    output: Optional[List[Argument]] = field(default_factory=EmptyList)
```

The fully AST python code is in the `vsc_ast.py`.

**Reading AST from YAML file**

```python
from vsc.model import vsc_parser
service = vsc_parser.get_ast_from_file('input.yaml')
```

**Building AST in the code**

```python
from vsc.model import vsc_ast
service = vsc_ast.AST(name='test', description='test', major_version=1, minor_version=0)
```

**Additional changes**

The PR contains additional changes. During the work on the dataclasses AST, the `vsc-tools` python module structure has been improved. Now we have `vsc` module with `generators`, `model`, `scripts`, `templates` submodules. `vsc_ast.py` contains AST specification implemented using python dataclasses. `vsc_generator.py` is responsible for the code generation. `vsc_parser.py` provides API to load and parse  YAML files into AST classes. 

```
├── LICENSE
├── Pipfile
├── Pipfile.lock
├── README.md
├── requirements.txt
├── setup.py
├── tests
│   ├── gen_test.py
│   ├── __init__.py
│   ├── input.yaml
│   ├── result
│   └── template
└── vsc
    ├── generators
    │   └── __init__.py
    ├── __init__.py
    ├── model
    │   ├── __init__.py
    │   ├── vsc_ast.py
    │   ├── vsc_generator.py
    │   └── vsc_parser.py
    ├── scripts
    │   ├── generator.py
    │   └── __init__.py
    └── templates
        ├── Argument-simple_doc.html
        ├── AST-simple_doc.tpl
        ├── dtdl.md
        ├── dtdl.tpl
        ├── Error-simple_doc.html
        ├── __init__.py
        ├── Member-simple_doc.html
        ├── Namespace-simple_doc.html
        ├── protobuf.md
        ├── protobuf.tpl
        ├── sds-bamm-aspect-model.md
        ├── sds-bamm-aspect-model.tpl
        ├── sds-bamm-macros.tpl
        ├── Service-simple_doc.html
        ├── simple_overview.tpl
        └── test.tpl
```

We added `setup.py` to simplify the installation of the module using `pip` or `python3 setup.py install` (`python3 setup.py develop`). 

**Installation from git**

```bash
pip3 install git+https://github.com/COVESA/vsc-tools.git
```

**Command line tools**

Newly added `setup.py` will install a command line tool. The tool called `vscgen` can be used to generate the code using default templates or custom templates.

```bash
vscgen <path-to-yaml-file> <template-file-name>
```
The list of templates can be found in the `README.md`. 

Mikhail Tsukerman <mikhail.tsukerman@mbition.io> on behalf of MBiTION GmbH. 

https://github.com/mercedes-benz/daimler-foss/blob/master/PROVIDER_INFORMATION.md

### CODE REVIEW

**Comment by gunnarx** [comment]
First I cherry-picked and pushed e8ec7ee because it seems to be needed for the GH actions to work now.
 (I found this out in PR #31).  

But now a new failure.

@miketsukerman Could you take a look at the buildtest failure now?  It looks like the import statement fails.  
Did you run gen_test.py manually (or run pytest?) because I can't make it work any directory but I only did a few quick tests.

Some hints:  Before your changes in f0d886a I had a bit of a hack to set the search path before import:  (see [here](https://github.com/COVESA/vsc-tools/blob/90b70c7274d508033157acc73981730829d14608/tests/gen_test.py#L13)
Maybe you can find a better way and still make it work?

**Comment by miketsukerman** [comment]
> First I cherry-picked and pushed [e8ec7ee](https://github.com/COVESA/vsc-tools/commit/e8ec7eec9d225fd01622cf92ed691ca9692ef653) because it seems to be needed for the GH actions to work now. (I found this out in PR #31).
> 
> But now a new failure.
> 
> @miketsukerman Could you take a look at the buildtest failure now? It looks like the import statement fails. Did you run gen_test.py manually (or run pytest?) because I can't make it work from any directory but I only did a few quick tests.
> 
> Some hints: Before your changes in [f0d886a](https://github.com/COVESA/vsc-tools/commit/f0d886af860a02b071b622281e7400403ecf8429) I had a bit of a hack to set the search path _before_ import: (see [HERE](https://github.com/COVESA/vsc-tools/blob/90b70c7274d508033157acc73981730829d14608/tests/gen_test.py#L13)) Maybe you can find a better way and still make it work?

I run the test from pycharm with dedicated python virtual environment. 

```python
# This is maybe not ideal way but seems efficient enough
from pathlib import Path
import sys
proj_root = Path(__file__).parents[1]
sys.path.append(str(proj_root) + "/model")
```
Should not be needed in general. I was asking myself why this repository doesn't have `setup.py`? I think if `setup.py` would be added should eliminate the problem.

**Comment by miketsukerman** [comment]
@gunnarx should pipeline start automatically?

**Comment by gunnarx** [comment]
@miketsukerman In addition to what we talked about (not parsing some of the tree correctly) I saw today that it is missing 
Enumerations.  [link to spec](https://github.com/COVESA/vehicle_service_catalog/blob/master/syntax.md#namespace-list-object-enumerations)

File: Pipfile [hunk 1]
```diff
@@ -13,3 +13,7 @@ pytest = ">=2.7.2"
 
 [requires]
 python_version = "3.10"
+
+[packages.e1839a8]
```
Comment (erikbosch) [inline Pipfile:17]: I am not a pip expert, but previously for VSS I got the impression that it is good practice to always have a Pipfile.lock as well as the Pipfile to "lock" the build to specific versions. Or do we see that as not needed here?

File: tests/gen_test.py [hunk 2]
```diff
@@ -3,37 +3,52 @@
...
 
-def test_gen():
-    # The files named 'input', 'template' and 'result' are in the tests directory
-    ast = vsc_parser.get_ast_from_file('input')
 
-    with open("template", "r") as templatefile:
-        generated = vsc_generator._gen_with_text_template(ast, templatefile.read())
+def test_gen(tmp_path):
+    # The files named 'input.yaml', 'template' and 'result' are in the tests directory
+    ast_root = vsc_parser.get_ast_from_file(os.path.join(TestPath,'input_old.yaml'))
```
Comment (erikbosch) [inline tests/gen_test.py:18]: Is this a test to support the "old" syntax (without explicit "service"). Do we intend to support both or is this just a temporary test?

File: Pipfile (hunk 1 above)
Comment (miketsukerman) [reply Pipfile:17]: I will bring it back later. There were issues with pipenv on Linux. During the experiments I have removed it.

File: tests/gen_test.py (hunk 2 above)
Comment (miketsukerman) [reply tests/gen_test.py:18]: We are in the process of migration and I've kept both. But, in the end, only one syntax will be left.

**Comment by gunnarx** [comment]
Yes these latest commits are similar changes I had to make…. Also, we already chatted about this but just to document:  I think the updated parsing function will instantiate all parts immediately so the post init functions here won’t  be necessary.  I will find some time later tomorrow to compare and merge it all once again, and then we will see if it works as expected.

File: tests/gen_test.py (hunk 2 above)
Comment (miketsukerman) [reply tests/gen_test.py]: @erikbosch I have reverted most of the changes I've made with "syntax". Some minor things are still there, like `input` instead of `in`, etc., but it is not avoidable.

**Comment by miketsukerman** [comment]
> Yes these latest commits are similar changes I had to make…. Also, we already chatted about this but just to document: I think the updated parsing function will instantiate all parts immediately so the post init functions here won’t be necessary. I will find some time later tomorrow to compare and merge it all once again, and then we will see if it works as expected.

yes, that would be great. If it works with templates we are good to go I guess.

File: tests/gen_test.py (hunk 2 above)
Comment (gunnarx) [reply tests/gen_test.py]: @erikbosch Honestly it's all a bit of a mess at the moment!  While merging and updating some of my code on top of this I noticed we had a mix of proposed syntax changes from @miketsukerman, a few from myself, the templates, the specification, different implementations in progress, unit-tests, examples files in VSC repo - all had various levels of differences.  To get a stable situation I have also rolled back a lot to a syntax that I don't feel is the final/best one.
But I'd rather we start from there and change incrementally.  From now on, let's try to make sure all PRs will modify parts in sync. (This is still Draft PR of course, but nonetheless).

Comment (miketsukerman) [reply tests/gen_test.py]: @gunnarx let's call it `refactoring` :)

File: Pipfile (hunk 1 above)
Comment (miketsukerman) [reply Pipfile]: @erikbosch I have reverted `Pipenv.lock` file back, please check and resolve the thread.

**Comment by gunnarx** [comment]
@miketsukerman We can't work like this...  There's no way to reasonably merge changes together when you have commits like this recent push.  It was a problem before as well that things were not separated the right way, and the squashing together of commits is rarely the solution to bad commits - it is likely to create even worse commits.   

Too many of these commits contain too many changes that are not related to each other.  The effect of this is that they can't be used separately or effectively combined with other changes without lots of merge conflicts. (b6fa1afcabee5d85bc9c15665884264f8fec4dfc is only one example).

Basically, I wonder why you didn't continue the work on top of my changes?  Do you think it's better to just do this on your own?

There are even deeper issues here that I can explain separately, but before I go too far - is all this just because you need some help to know how to separate things in git, or are you kind of just developing without regard for others?

File: vsc/model/vsc_generator.py [hunk 3]
```diff
@@ -58,7 +58,7 @@ def get_template(filename):
     return jinja_env.get_template(filename)
 
 # gen() function to be called from templates
-def gen(node : AST, template_file = None):
+def gen(node : Namespace, template_file = None):
```
Comment (gunnarx) [inline vsc/model/vsc_generator.py:61]: 1. The idea to have the type as AST came from the original implementation where all node-types were sub-classes inheriting from the AST class.  Thus the type of node "is-a" AST.  But that class hierarchy does not exist anymore.  

2. Even if the top-level node will be Namespace, the purpose of the gen() function is that it can accept **any** node type.

So... in this version where the class AST does not exist, I would guess that it is just as good to not give a type hint at all.

File: vsc/model/vsc_generator.py [hunk 4]
```diff
@@ -12,14 +12,14 @@
 
 # This performs the following related functions:
 
-# 1. As input.yaml we expect an AST as provided by the parser module
+# 1. As input.yaml we expect an Namespace as provided by the parser module
```
Comment (gunnarx) [inline vsc/model/vsc_generator.py:15]: I would still say that the whole thing that is provided into the generator is an Abstract Syntax Tree, even if in the current syntax definition the top level node type is Namespace.

I guess in this comment the word AST could be considered the concrete class name, but it could also be considered the "concept" of providing an AST from the parser.  I would personally keep the name AST.

File: vsc/model/vsc_generator.py (hunk 3 above)
Comment (miketsukerman) [reply vsc/model/vsc_generator.py:61]: 1. Once you mentioned that you wanted to make everything according to the specification and in the specification, there is no root element except `Namespace`. No problem, I will fix it later today and bring `AST` back.

2. Hm, for this purpose in python there is a generic type, something like that would be 
```python
def gen(node: Generic[T], template_file = None):

```

File: vsc/model/vsc_generator.py (hunk 4 above)
Comment (miketsukerman) [reply vsc/model/vsc_generator.py]: I've return AST node back.

File: vsc/model/vsc_generator.py (hunk 3 above)
Comment (miketsukerman) [reply vsc/model/vsc_generator.py]: @gunnarx done, you can check.

Comment (miketsukerman) [reply vsc/model/vsc_generator.py]: well, just thought

```python
def gen(node: Any, template_file = None):
...
```
would work as well

**Comment by miketsukerman** [comment]
@gunnarx do we need code below in the `vsc_parser.py`?

```python
# If run as a script, generate a single YAML file using a single root template
if __name__ == "__main__":
    if not len(sys.argv) == 3:
        usage()
    ast = vsc_ast.read_ast_from_yaml_file(sys.argv[1])
    templatename = sys.argv[2]
    print(gen(ast, templatename))

```

**Comment by gunnarx** [comment]
> 1.  Do we need code below in the vsc_parser.py? 
> We have vscgen instead.

OK, remove it in **vsc_generate**.py we no longer need it.

I have been meaning to clean up the modules also, so I agree with it generally, but I wanted changes to slow down a bit first.

> 2. I would like to move some code from the vsc_generator.py to templates module

Should be OK.  As long as it's possible to redefine some of those variables if a custom generator is written.
Especially `default_templates`, which has specific templates listed for a specific case.

So basically    `$ <vscgen/vsc_generator> <input-file> <template>`  is fine for some simple cases, but I think in some of the cases there will be a separate generator written (which imports the modules and makes a few settings before calling the generation function).

I have some ideas for how to organize it, but happy to look at what you might propose if it's a separate commit.

I think we might also consider slowing down a bit, deciding what to merge from here and start again.  It's easier to take a step at a time.

File: vsc/model/vsc_generator.py (hunk 3 above)
Comment (miketsukerman) [reply vsc/model/vsc_generator.py]: > OK... point 1 was just information, not really a change request. As I said the class hierarchy does not exist now, so therefore there is (was) no AST parent type.

I know. I was not sure completely, waiting for your feedback.

**Comment by miketsukerman** [comment]
> > 1. Do we need code below in the vsc_parser.py?
> >    We have vscgen instead.
> 
> OK, remove it in **vsc_generate**.py we no longer need it.
> 

Okay. 

> I have been meaning to clean up the modules also, so I agree with it generally, but I wanted changes to slow down a bit first.
> 
> > 2. I would like to move some code from the vsc_generator.py to templates module
> 
> Should be OK. As long as it's possible to redefine some of those variables if a custom generator is written. Especially `default_templates`, which has specific templates listed for a specific case.
> 

The point about redefinition of the `default_templates` I am not sure if I got fully. 

> So basically `$ <vscgen/vsc_generator> <input-file> <template>` is fine for some simple cases, but I think in some of the cases there will be a separate generator written (which imports the modules and makes a few settings before calling the generation function).
> 
> I have some ideas for how to organize it, but happy to look at what you might propose if it's a separate commit.
> 
> I think we might also consider slowing down a bit, deciding what to merge from here and start again. It's easier to take a step at a time.

The "custom" generator would most likely use its own wrappers over Jinja, cause anyway, a custom generator needs custom templates and so on. If I were a person who would like to write a generator for VSC, I would use only `vsc_ast` and `vsc_parser`.

**Comment by miketsukerman** [comment]
> I think we might also consider slowing down a bit, deciding what to merge from here and start again. It's easier to take a step at a time.

Let's do it ;)

File: vsc/model/vsc_ast.py [hunk 5]
```diff
@@ -24,31 +24,31 @@ class Error:
 class Method:
     name: str
     description: Optional[str] = None
-    error: Optional[List[Error]] = None
-    input: Optional[List[Argument]] = None
-    output: Optional[List[Argument]] = None
+    error: Optional[List[Error]] = field(default_factory=lambda: [])
```
[the remark this answers is not in the document]
Comment (miketsukerman) [reply vsc/model/vsc_ast.py:26]: https://docs.python.org/3/library/dataclasses.html#mutable-default-values

```bash
tests/gen_test.py:7: in <module>
    from vsc.model import vsc_ast, vsc_parser, vsc_generator
vsc/model/vsc_ast.py:98: in <module>
    class Namespace:
/usr/lib/python3.10/dataclasses.py:1185: in dataclass
    return wrap(cls)
/usr/lib/python3.10/dataclasses.py:1176: in wrap
    return _process_class(cls, init, repr, eq, order, unsafe_hash,
/usr/lib/python3.10/dataclasses.py:956: in _process_class
    cls_fields.append(_get_field(cls, name, type, kw_only))
/usr/lib/python3.10/dataclasses.py:813: in _get_field
    raise ValueError(f'mutable default {type(f.default)} for field '
E   ValueError: mutable default <class 'list'> for field namespaces is not allowed: use default_factory

```

[the remark this answers is not in the document]
Comment (miketsukerman) [reply vsc/model/vsc_ast.py:26]: For `description: Optional[str] = ""` the construct `description: str = str()` seems to be working

**Comment by gunnarx** [comment]
> The point about redefinition of the default_templates I am not sure if I got fully.

The idea of the `gen()` function is that you can pass any node type to it.  It can then pick a template that was predefined for that node type.  In other words, for complex generation you don't need to put everyting into only one jinja template file - it can be divided down into smaller files that handle different sub-nodes using different templates.  
When using this feature, the predefined `default_templates` will be unique to a particular code generation, and so it is likely that you need to write your own python file that reuses the parser/generator modules.   In that python file you would define `default_templates` and some other jinja settings the way you like them, then delegate most of the work to the parser/generator modules.   `vscgen` is useful for simple tests and single-template generation, but for more complex ones a new python main file can be added.  

In other words, we should probably make `default_templates` empty when using the `vscgen` entry point, because it is now set to "simple-overview" type templates, and it doesn't really make sense to have anything like that hardcoded.

I can put some work into that when I feel I have a stable starting point.

**Comment by erikbosch** [comment]
Concerning default templates (and templates in general) I found when we added the DTDL/protobuf/BAMM templates some time ago that many of them are not functional (see overview in https://github.com/COVESA/vsc-tools#existing-toolstemplates).

For now (for this PR) I assume it is OK that some templates are broken, but if so we should have a follow up task to either fix them so they work with our example service (i.e. the seat service) or remove them. If needed I can help refactoring/testing the templates contributed by Bosch so they are still functional when being used with the seat service. But that would happen first when this PR is stable and we have an example template (e.g. simple overview) working using the new AST.

**Comment by gunnarx** [comment]
> Concerning default templates (and templates in general) I found when we added the DTDL/protobuf/BAMM templates some time ago that many of them are not functional (see overview in https://github.com/COVESA/vsc-tools#existing-toolstemplates).

Some are definitely updated now.  But in any case I think we should look over them after cleaning up the usage of default_templates as per the previous comment.

> but if so we should have a follow up task to either fix them so they work with our example service (i.e. the seat service) or remove them. 

I don't disagree.

Thanks for the offer to update.  In general I think it should be the responsibility of this (any) PR to also update templates if it changes the AST, but we have not been perfect in the past (and maybe not this time either).

**Comment by gunnarx** [comment]
@erikbosch Oh I did not realize we had that table in the README... OK so that should also be looked at.

**Comment by miketsukerman** [comment]
> > The point about redefinition of the default_templates I am not sure if I got fully.
> 
> No problem. The idea of the `gen()` function is that you can pass any node type to it. It can then pick a template that was predefined for that node type. In other words, for complex generation you don't need to put everything into only one jinja template file - it can be divided down into smaller files that handle different sub-nodes using different templates. (See Namespace-simple_doc.html template for example - it calls gen() on most of the child nodes, and then the templates in `default_templates` are used to define the output).
> 
> When using this feature, the predefined `default_templates` will be unique to a particular code generation, and so it is likely that you need to write your own python file that reuses the parser/generator modules. In that python file you would define `default_templates` and other jinja settings the way you like them for this particular code-generation and then delegate most of the work to the parser/generator modules we have written.
> 

Thank you for the explanation. I have assumed something similar to what you just described while I was reading the code. Maybe some small concern from my side. Would be better, at least in my opinion, to avoid the use of variables like `default_templates = {}` that modify the code behavior globally. I do not insist, of course, on the final path we would decide to take, but I can propose to encapsulate the code into an object, something like:

```python
class VscGenerator:
        def __init__(self, default_templates: List[str]):
              self.templates = default_templates;
       
       def gen(self, ast: Any):
            ....
```

Then we can pass here a list of templates defined by vsc-tools `default_templates`. Or any other list created by someone else.  

> In other words, `vscgen` is useful for simple tests and single-template generation, but for more complex ones a new python main file (a new executable program) can be added.
> 
> So I think it means that we should probably make `default_templates` empty when using the `vscgen` entry point, because it is now set to "simple-overview" type templates, and it doesn't really make sense to have anything like that hardcoded.
> 
> I can put some work into that when I feel I have a stable starting point.

Hmm, `vscgen` works with multiple templates, like `*-simple_doc.html` as well. If you call it without a template name it will work using `gen()` function behavior picking up templates from `vsc/templates/` folder using AST node names. But, of course, we can have multiple python executable files with different behavior.

In theory, we can pass the list of templates via the command line as well to remove the need for the `default_templates`. Or, even just a folder with templates, if the templates would have used a special naming convention, like `<Node>-<template_type>.<template_ext>`.

**Comment by gunnarx** [comment]
> Would be better, at least in my opinion, to avoid the use of variables like default_templates = {} that modify the code behavior globally. 

Yeah that's fine, and I anticipated/planned similar changes later on if needed.  It doesn't cause any issue for our current simple generator programs because within context of that program we basically want to set the global behavior!!  But if more advanced or different designs appear later it may need to change, so if you want to do it now that's fine.

> I do not insist, of course, on the final path we would decide to take, but I can propose to encapsulate the code into an object, >something like:
>> 
>> class VscGenerator:
>>         def __init__(self, default_templates: List[str]):
>>               self.templates = default_templates;
>>        
>>        def gen(self, ast: Any):
>>             ....

I think this anticipates a program that needs to instantiate more than one instance of the generator, and maybe such programs will be developed, but later I guess.  

It's of course not a bad design change you're suggesting so if you want to do it now, just go ahead.  

One thing, can you still pass vsc_generator_instance.gen()` into the jinja environment so that it looks like a global function (or at least one imported into the same namespace) from within the template as it is now.  I at least like that it is immediately available to the template code.

**Comment by gunnarx** [comment]
>Hmm, vscgen works with multiple templates, like *-simple_doc.html as well. If you call it without a template name it will work >using gen() function behavior picking up templates from vsc/templates/ folder using AST node names. 

Yeah it works I suppose because default_templates is "hard coded" to point to the simple_doc templates so it works only for those, and that doesn't make sense (and those are just some simple test templatest anyway, nothing useful).

My point was, vscgen is not a _generally useful_ executable even if you pass it a template argument, because it is also controlled by the default_templates which are hard-coded in the file right now.  

> In theory, we can pass the list of templates via the command line as well to remove the need for the default_templates.

Yeah I also have thought of that but it feels very clumsy to list many templates like that.   I never liked the idea.

> Or, even just a folder with templates, if the templates would have used a special naming convention, like <Node>-<template_type>.<template_ext>.](url)

That feels like a much better idea I think.  I'd be more positive to a command line switch that gets only one argument, such as if we pass it a template-set, a directory, where both the top node and any other node templates are located.

Or it could accept either a file (single template only, gets the top node as input) or a dir (behavior like default_templates today).

**Comment by miketsukerman** [comment]
> It's of course not a bad design change you're suggesting so if you want to do it now, just go ahead.

Well, it was more a suggestion for the next changes. At the moment, change set in the current MR is complete. In general, could be split into two PR's - "module structure refactoring" and "dataclasses for AST". 

>   That feels like a much better idea I think. I'd be more positive to a command line switch that gets only one argument, such as > if we pass it a template-set, a directory, where both the top node and any other node templates are located.
> Or it could accept either a file (single template only, gets the top node as input) or a dir (behavior like default_templates today).

Templates could also be a python module on its own. In theory, one can think of the small python wrapper over templates (if we will come up with some API) and `vscgen` then would just load them by name. 

```bash
pip3 install vsc-templates-someip-cpp
vscgen seats.yaml vsc-templates-someip-cpp
```

Where `vscgen` would then try to import `vsc-templates-someip-cpp` internally in the runtime to get the list of templates and run the generator.

File: vsc/model/vsc_ast.py [hunk 6]
```diff
@@ -23,63 +26,63 @@ class Error:
...
+    input: List[Argument] = field(default_factory=EmptyList)
+    output: List[Argument] = field(default_factory=EmptyList)
 
 
 @dataclass
 class Event:
     name: str
-    description: Optional[str] = None
-    input: Optional[List[Argument]] = field(default_factory=lambda: [])
+    description: str = None
```
Comment (gunnarx) [inline vsc/model/vsc_ast.py:37]: Some have `str = None`, some have `str = str()`
Also here the description is an _optional_ thing.

File: vsc/model/vsc_ast.py [hunk 7]
```diff
@@ -23,63 +26,63 @@ class Error:
 class Method:
     name: str
     description: Optional[str] = None
-    error: Optional[List[Error]] = field(default_factory=lambda: [])
-    input: Optional[List[Argument]] = field(default_factory=lambda: [])
-    output: Optional[List[Argument]] = field(default_factory=lambda: [])
+    error: List[Error] = field(default_factory=EmptyList)
```
Comment (gunnarx) [inline vsc/model/vsc_ast.py:29]: Makes sense to factor out the lambda.  Reduced repetition of code (even if very small).

But if possible I think we should keep **Optional** in the type hint, because it more clearly/explicitly shows that this item is optional.

Comment (miketsukerman) [reply vsc/model/vsc_ast.py]: I brought back the `Optional[...]` , check in case I have missed some of them.

File: vsc/model/vsc_ast.py (hunk 6 above)
Comment (miketsukerman) [reply vsc/model/vsc_ast.py]: When you have a field `description: Optional[str] = None` in the template without checking the current value, you would have `None` as a string. So, for some of the fields, like description, I added `str()` to not have None in the template output. 
This is also a more philosophical question.

File: vsc/model/vsc_ast.py [hunk 8]
```diff
@@ -96,21 +96,60 @@ class Include:
...
+    namespaces:
+      - name: seats
+        major_version: 1
+        minor_version: 3
+        description: Seat interface and datatypes.
+    ```
+    """
+
     name: str
+    """ Specifies the name of the namespace. """
```
Comment (miketsukerman) [inline vsc/model/vsc_ast.py:121]: @gunnarx how about this style of comments? Just curious about your opinion.

Comment (gunnarx) [reply vsc/model/vsc_ast.py:121]: In general it makes sense to me if we use it as a (single) source for the documentation and language definition.  
Otherwise it feels like extra stuff to maintain, (and in many cases giving  information that you already know.  I mean, we try to make sure the name of each thing understandable already.   A variable _name_ inside a class called _Namespace_, it's quite obvious that it  _Specifies the name of the namespace_. ".)  
Even despite this, I guess I feel it's OK to write something similar, _iff_ we put it in a single place and it is used as the documentation and official language definition.   

(In many cases I think normally you have to write either nothing, or you have to write a bit more to explain what something really is).

But on the topic of putting this information in one single place - it should work just fine.  The idea shown in #37 could extract the `__doc__` parameter to get the """comments""", and use that to output docs. 

I still think the group would better discuss and agree on some fundamental things first (https://github.com/COVESA/vehicle_service_catalog/pull/41), and then when we have a basic agreement, put it into code.

Also, if we put all of it in the code, it moves the language definition from https://github.com/COVESA/vehicle_service_catalog/ to **vsc-tools**, so that's another decision to agree on in that case.

**Comment by gunnarx** [comment]
Meeting 19 September: OK to merge after adding appropriate license-headers and signoffs.

**Comment by miketsukerman** [comment]
> Meeting 19 September: OK to merge after adding appropriate license-headers and signoffs.

Updated PR description with short summary. Added license headers etc.
